incubator/hnc/internal/validators/namespace.go

```diff
@@ -137,7 +137,17 @@ func (v *Namespace) cannotDeleteSubnamespace(req *nsRequest) admission.Response
 }
 
 func (v *Namespace) illegalCascadingDeletion(ns *forest.Namespace) admission.Response {
-	if ns.AllowsCascadingDeletion() || len(ns.ChildNames()) == 0 {
+	if ns.AllowsCascadingDeletion() {
+		return allow("")
+	}
+
+	hasSubNs := false
+	for _, cnm := range ns.ChildNames() {
+		if v.Forest.Get(cnm).IsSub {
+			hasSubNs = true
+		}
+	}
+	if !hasSubNs {
 		return allow("")
 	}
```

Comment:
This could be even simpler:

```
for _, cnm := range ns.ChildNames() {
  if v.Forest.Get(cnm).IsSub {
    msg := "This namespaces contains subnamespaces. Please remove all subnamespaces before deleting this namespace, or set 'allowCascadingDeletion' to delete them automatically."
    return deny(metav1.StatusReasonForbidden, msg)
  }
}

return allow("no subnamespaces found")
```

Reply:
+1

Reply:
done